File: packages/core-dotnet/ClientData.cs
using System.Collections.Generic;
using System.Text;
using System.Text.Json;

namespace OpenPasskey.Core;

internal static class ClientData
{
    public static byte[] Verify(string clientDataJSONB64, string expectedType, string expectedChallenge, string expectedOrigin, IReadOnlyList<string>? additionalOrigins = null)
    {
        byte[] raw = Base64Url.Decode(clientDataJSONB64);
        string text = Encoding.UTF8.GetString(raw);
        using JsonDocument doc = JsonDocument.Parse(text);
        JsonElement root = doc.RootElement;

        string type = root.GetProperty("type").GetString()!;
        string challenge = root.GetProperty("challenge").GetString()!;
        string origin = root.GetProperty("origin").GetString()!;

        if (type != expectedType)
            throw new WebAuthnException("type_mismatch");
        if (challenge != expectedChallenge)
            throw new WebAuthnException("challenge_mismatch");
        if (origin != expectedOrigin && (additionalOrigins == null || !additionalOrigins.Contains(origin)))
            throw new WebAuthnException("origin_mismatch");

        if (root.TryGetProperty("tokenBinding", out JsonElement tb))
        {
            if (tb.TryGetProperty("status", out JsonElement status) && status.GetString() == "present")
                throw new WebAuthnException("token_binding_unsupported");
        }

        return raw;
    }
}

File: packages/core-java/src/main/java/com/openpasskey/core/ClientData.java
package com.openpasskey.core;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

/**
 * Client data JSON verification.
 */
public final class ClientData {
    private ClientData() {}

    private static final ObjectMapper JSON_MAPPER = new ObjectMapper();

    /**
     * Verifies the clientDataJSON and returns the raw decoded bytes.
     */
    public static byte[] verify(String clientDataJSONB64, String expectedType,
                                String expectedChallenge, String expectedOrigin)
            throws WebAuthnException {
        byte[] raw = Base64Url.decode(clientDataJSONB64);
        JsonNode json;
        try {
            json = JSON_MAPPER.readTree(raw);
        } catch (Exception e) {
            throw new WebAuthnException("invalid_client_data", "Failed to parse clientDataJSON");
        }

        // Verify type
        String type = json.has("type") ? json.get("type").asText() : null;
        if (!expectedType.equals(type)) {
            throw new WebAuthnException("type_mismatch",
                    "Expected type '" + expectedType + "' but got '" + type + "'");
        }

        // Verify challenge
        String challenge = json.has("challenge") ? json.get("challenge").asText() : null;
        if (!expectedChallenge.equals(challenge)) {
            throw new WebAuthnException("challenge_mismatch",
                    "Challenge does not match expected value");
        }

        // Verify origin
        String origin = json.has("origin") ? json.get("origin").asText() : null;
        if (!expectedOrigin.equals(origin)) {
            throw new WebAuthnException("origin_mismatch",
                    "Expected origin '" + expectedOrigin + "' but got '" + origin + "'");
        }

        // Check token binding
        if (json.has("tokenBinding")) {
            JsonNode tb = json.get("tokenBinding");
            if (tb.has("status") && "present".equals(tb.get("status").asText())) {
                throw new WebAuthnException("token_binding_unsupported",
                        "Token binding status 'present' is not supported");
            }
        }

        return raw;
    }
}
